FactSet/factset_fund_data.R

```diff
@@ -0,0 +1,198 @@
+
+# Sanity Check: all funds for which a factset_fund_id exists has complete
+# coverage on reported_mv (complete in the sense that every security has an
+# associated market value, doesn't necessarily mean each fund has comprehensive
+# coverage of all securities)
+data %>%
+  mutate(factset_fund_id_exists = !is.na(factset_fund_id)) %>%
+  filter(factset_fund_id_exists) %>%
+  mutate(mv_exists = !is.na(reported_mv)) %>%
+  group_by(factset_fund_id_exists) %>%
+  summarize(isin_count = n())
+
+# Using the following two issues and asset maps, we can get info on what
+# asset_class is associated with each holding
+ref_v2_issue_type_map <- tbl(factset_db, "ref_v2_issue_type_map")
+ref_v2_asset_class_map <- tbl(factset_db, "ref_v2_asset_class_map")
+
+issue_type_code_asset_class_desc <- ref_v2_issue_type_map %>%
+  left_join(ref_v2_asset_class_map, by = "asset_class_code") %>%
+  select(issue_type_code, asset_class_desc)
+
+fsym_id_asset_type <- tbl(factset_db, "own_v5_own_sec_coverage") %>%
+  left_join(
+    issue_type_code_asset_class_desc,
+    by = c("issue_type" = "issue_type_code")
+    ) %>%
+  select(fsym_id, asset_type = asset_class_desc) %>%
+  collect()
+
+# connect asset type to fund data
+data <- data %>%
+  left_join(fsym_id_asset_type) %>%
+  select(fund_isin, reported_mv, asset_type, fsym_id)
+
+# connect ISINs to the fsym_ids containing each
+data <- data %>%
+  left_join(sym_isin) %>%
+  select(fund_isin, holding_isin = isin, reported_mv, asset_type)
+
+# It's important to note that each fund may have different types of holdings:
+# Equity, Fixed Income, Private Equity, Cash, Other
+# We don't expect Private Equity, Cash or Other to necessarily have corresponding ISINs
+# However, the Fixed Income class could be an issue.
+
+# Let's look at coverage by asset_type
+data %>%
+  filter(!is.na(asset_type)) %>%
+  mutate(holding_isin_exists = !is.na(holding_isin)) %>%
+  group_by(fund_isin, asset_type, holding_isin_exists) %>%
+  summarise(reported_mv = sum(reported_mv)) %>%
+  group_by(fund_isin) %>%
+  mutate(total_mv = sum(reported_mv)) %>%
+  group_by(fund_isin, holding_isin_exists) %>%
+  mutate(percent_coverage = reported_mv / total_mv)
+
+# we can also look at how many funds have above 50% coverage by AuM
+coverage <- data %>%
+  filter(!is.na(asset_type)) %>%
+  mutate(has_isin = !is.na(holding_isin)) %>%
+  group_by(fund_isin, has_isin) %>%
+  summarise(reported_mv = sum(reported_mv)) %>%
+  group_by(fund_isin) %>%
+  mutate(coverage = reported_mv / sum(reported_mv)) %>%
+  filter(has_isin) %>%
+  select(fund_isin, coverage)
+
+# 617 funds by ISIN
+nrow(coverage)
+
+# 430 funds with above 50% coverage
+coverage %>%
+  filter(coverage > 0.5) %>%
+  nrow()
+
+# format the final FactSet data that can be a place in for fund_data
+factset_fund_data <- data %>%
+  group_by(fund_isin) %>%
+  mutate(total_mv = sum(reported_mv)) %>%
+  group_by(fund_isin, holding_isin) %>%
+  summarise(isin_weight = sum(reported_mv)/sum(total_mv)) %>%
```

Comment:
just to clarify, we are adding together the market values related to each fund ID to get the total and then dividing each holding's market value from the same fund with this total? 

The definition of reported market value is this according to factset: Market value held by the fund as reported in the source, represented in USD
so i am still confused as to why the reported market value would change per row... why isn't there a 'holdings value' to show the value of the holdings within the fund? (using the same image as before) 
![image](https://user-images.githubusercontent.com/55149721/168114855-056fa61a-9b67-4c12-bba8-6340011ec427.png)

Reply:
Market value HELD by fund = market value of the holding. 

In the table above: 
- `factset_fund_id` indicates one single fund (a portfolio of holdings)
- `fsym_id` indicates one single holding
- `reported_mv` indicates the reported market value of that holding. 

`reported_holding` and `reported_mv` are similar values, the former is the "share amount" of the holding, the latter is the "market value" of the holding 

To get the market value of the fund, you just have to aggregate the market value of all holdings.

Reply:
wow. I can't believe it took me this long to understand this. Thanks Jackson!

Reply:
> To get the market value of the fund, you just have to aggregate the market value of all holdings.


assuming you have/know *all* holdings

Reply:
soo waiittt what if some of the holdings are cash and don't have an isin, then u don't have info on ALL the holdings right? can you still do this aggregation and add the caveat that "non-isin-containing-financial-vehicles" are not included in this aggregation?

Reply:
that is why there are often holdings (at least in the Lipper data) that have `"Cash"` or `"MissingWeight"` in the `isin` column

Reply:
and that is why we typically, or should aggregate all non-ivestment vehicle holdings into one `"Other"` or `"MissingWeight"` row

Reply:
however, it should be a concern if we believe the data *does not* have *all* holdings, including non-investment vehicles, i.e. no isin, holdings.... or if we cannot validate or infer what is missing otherwise

Reply:
or if we know from elsewhere that the total value of the fund is `100` and we have

| fund_isin  | holding_isin | value |
| ------------- | ------------- | ------------- |
| 1286742  | 9028323  | 20 |
| 1286742  | 4359823  | 70 |

we can infer the rest as

| fund_isin  | holding_isin | weight |
| ------------- | ------------- | ------------- |
| 1286742  | 9028323  | 0.2 |
| 1286742  | 4359823  | 0.7 
| 1286742  | Unknown  | 0.1 |

Reply:
or if we know from elsewhere that the total value of the fund is 100 and we have

| fund_isin  | holding_isin | value |
| ------------- | ------------- | ------------- |
| 1286742  | 9028323  | 20 |
| 1286742  | 4359823  | 40 |
| 1286742  | Gold  | 10 |
| 1286742  | Silver  | 10 |
| 1286742  | Cash  | 10 |

we can infer

| fund_isin  | holding_isin | weight |
| ------------- | ------------- | ------------- |
| 1286742  | 9028323  | 0.2 |
| 1286742  | 4359823  | 0.4 | 
| 1286742  | Unknown  | 0.4 |

Reply:
there are many possibilities, but we need an end result that looks like

| fund_isin  | holding_isin | weight |
| ------------- | ------------- | ------------- |
| 1286742  | 9028323  | 0.2 |
| 1286742  | 4359823  | 0.4 | 
| 1286742  | Unknown  | 0.4 |

Reply:
but in our case, we have to calculate the weight and the total market value.. we may not necessarily know if whether we have all the holdings or not from the factset data so how can we make the inference..?

Reply:
and if "Unknown" is >= 0.5 (or some threshold) for any given fund, then we should consider whether or not we consider that a reliable source for that fund

Reply:
That is why I added the caveat "assuming you have/know all holdings". I didn't work on this, so I'm not sure how @jdhoffa got around that problem. But as I've demonstrated, there are many possible ways of getting around that (I only demonstrated a few).

Reply:
i get that but we don't know what the unknown would be since we have to calculate the total market value from individual holding information... like the total may not even be the full total if there is holding information that is missing and we don't have an external source to confirm what the actual total value of the fund is... does what i am saying make sense? so can we really be confident about our numbers?

Reply:
yes..... like I said, this is something that @jdhoffa should explain

Reply:
yea sorry, i didn't see ur other msg before😅. i was trying to formulate my thoughts 😅

Reply:
Ya so this could be an issue. The response from Ryan (FactSet) was that the total size of the fund is determined by `reported_mv`, which to my understanding, means aggregating the size of the fund, but there is no proof/ external validation that we are missing weights. 

That said, the FactSet data DOES at least take into account: "Equity", "Fixed Income", "Cash", "Other", etc. , so in principle, we know that most asset classes are covered, the issue would arise from lack of coverage. 

I did manually check the `isin_weight` of a handful of the randomly sampled funds between the `pacta-data/fund_data` and `factset_fund_data`, and found all the weights to be in the same range +/-0.1% or so, but granted this could still be a problem, and not entirely sure how to get around it.

Reply:
@AlexAxthelm FYI

Reply:
Ya, perhaps you can just elaborate on the last e-mail? 

The question being if there is a way to determine the total fund size other than aggregating `reported_mv`, as a way to cross-validate how much of the fund is covered